src/main/cpp/src/json_parser.hpp

```diff
@@ -0,0 +1,1786 @@
+
+/**
+ * JSON token enum
+ */
+enum class json_token {
+  // start token
+  INIT = 0,
+
+  // successfully parsed the whole JSON string
+  SUCCESS,
+
+  // get error when parsing JSON string
+  ERROR,
+
+  // '{'
+  START_OBJECT,
+
+  // '}'
+  END_OBJECT,
+
+  // '['
+  START_ARRAY,
+
+  // ']'
+  END_ARRAY,
+
+  // e.g.: key1 in {"key1" : "value1"}
+  FIELD_NAME,
+
+  // e.g.: value1 in {"key1" : "value1"}
+  VALUE_STRING,
+
+  // e.g.: 123 in {"key1" : 123}
+  VALUE_NUMBER_INT,
+
+  // e.g.: 1.25 in {"key1" : 1.25}
+  VALUE_NUMBER_FLOAT,
+
+  // e.g.: true in {"key1" : true}
+  VALUE_TRUE,
+
+  // e.g.: false in {"key1" : false}
+  VALUE_FALSE,
+
+  // e.g.: null in {"key1" : null}
+  VALUE_NULL
+
+};
+
+/**
+ * JSON parser, provides token by token parsing.
+ * Follow Jackson JSON format by default.
+ *
+ *
+ * For JSON format:
+ * Refer to https://www.json.org/json-en.html.
+ *
+ * Note: when setting `allow_single_quotes` or `allow_unescaped_control_chars`,
+ * then JSON format is not conventional.
+ *
+ * White space can only be 4 chars: ' ', '\n', '\r', '\t',
+ * Jackson does not allow other control chars as white spaces.
+ *
+ * Valid number examples:
+ *   0, 102, -0, -102, 0.3, -0.3
+ *   1e-5, 1E+5, 1e0, 1E0, 1.3e5
+ *   1e01 : allow leading zeor after 'e'
+ *
+ * Invalid number examples:
+ *   00, -00   Leading zeroes not allowed
+ *   infinity, +infinity, -infinity
+ *   1e, 1e+, 1e-, -1., 1.
+ *
+ * When `allow_single_quotes` is true:
+ *   Valid string examples:
+ *     "\'" , "\"" ,  '\'' , '\"' , '"' , "'"
+ *
+ *  When `allow_single_quotes` is false:
+ *   Invalid string examples:
+ *     "\'"
+ *
+ *  When `allow_unescaped_control_chars` is true:
+ *    Valid string: "asscii_control_chars"
+ *      here `asscii_control_chars` represents control chars which in Ascii code
+ * range: [0, 32)
+ *
+ *  When `allow_unescaped_control_chars` is false:
+ *    Invalid string: "asscii_control_chars"
+ *      here `asscii_control_chars` represents control chars which in Ascii code
+ * range: [0, 32)
+ *
+ */
+template <bool allow_single_quotes           = curr_allow_single_quotes,
+          bool allow_unescaped_control_chars = curr_allow_unescaped_control_chars,
+          int max_json_nesting_depth         = curr_max_json_nesting_depth,
+          int max_string_utf8_bytes          = curr_max_string_utf8_bytes,
+          int max_num_len                    = curr_max_num_len,
+          bool allow_tailing_sub_string      = curr_allow_tailing_sub_string>
+class json_parser {
+ public:
+  CUDF_HOST_DEVICE inline json_parser(char const* const _json_start_pos,
+                                      cudf::size_type const _json_len)
+    : json_start_pos(_json_start_pos),
+      json_end_pos(_json_start_pos + _json_len),
+      curr_pos(_json_start_pos)
+  {
+  }
+
+ private:
+  /**
+   * is current position EOF
+   */
+  CUDF_HOST_DEVICE inline bool eof(char const* pos) { return pos >= json_end_pos; }
```

Comment:
Also `CUDF_HOST_DEVICE`s need to be only `__device__` if possible.

Reply:
done.